diff --git a/fcping.c b/fcping.c
index f92516f..0dd8bd4 100644
--- a/fcping.c
+++ b/fcping.c
@@ -508,9 +508,9 @@ fp_find_hba(void)
 			fc_wwn_t wwn1;
 
 			wwn1 = fp_parse_wwn(fp_hba, "HBA", 2, 0);
-			wwn1 &= 0xffff000000000000;
+			wwn1 &= 0xffff000000000000LL;
 			wwn = fp_parse_wwn(&fp_hba[6], "HBA", 2, 0);
-			wwn &= 0x0000ffffffffffff;
+			wwn &= 0x0000ffffffffffffLL;
 			wwn |= wwn1;
 		} else if (strlen(fp_hba) == strlen("aa:bb:cc:dd:ee:ff")) {
 			wwn = fp_parse_wwn(fp_hba, "HBA", 2, 0);
